test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java

```diff
@@ -965,19 +965,28 @@ public long waitForDocs(final long numDocs, int maxWaitTime, TimeUnit maxWaitTim
                     lastKnownCount.set(count);
                 } catch (Exception e) { // count now acts like search and barfs if all shards failed...
                     logger.debug("failed to executed count", e);
-                    return false;
+                    throw e;
                 }
                 logger.debug("[{}] docs visible for search. waiting for [{}]", lastKnownCount.get(), numDocs);
             } else {
                 logger.debug("[{}] docs indexed. waiting for [{}]", lastKnownCount.get(), numDocs);
             }
-            return lastKnownCount.get() >= numDocs;
+
+            assertThat(lastKnownCount.get(), greaterThanOrEqualTo(numDocs));
         };
 
-        while (!awaitBusy(testDocs, maxWaitTime, maxWaitTimeUnit)) {
+        while (true) {
+            try {
+                assertBusy(testDocs, maxWaitTime, maxWaitTimeUnit);
```

Comment:
Perhaps we should just make the maxwaittime scale based on number of docs instead of doing several assert busy calls?

Reply:
Happy to do this but I could use some input. The main implementation of `waitForDocs` is only called by an override, which uses a max wait time of 90s. I looked at the usages of `waitForDocs`, and the possible range for `numDocs` is 0 to 2500. (There is one test that indexes 10000 docs, but it calls `waitForDocs` multiple times). At the moment, `waitForDocs` would continue happily so long as 1 document was indexed every 90s. So the question is, how long does it take to index a document?

Reply:
I think going with something conservative would be fine as a start. 200ms per doc? We can have use a hardcoded min value for the wait time to make sure it isn't too small if say we had 3 heavyweight documents being index.

Reply:
I went with 90s due to the comment in the method's override.

Reply:
90s seems like it won't give any guarantees. We will hit the test suite timeout of 20 mins long before we hit 900k seconds for 10k docs.

Reply:
Sorry, I should have been clearer - 90s is the minimum wait time, due to the comment:

> indexing threads can wait for up to ~1m before retrying when they first try to index into a shard which is not STARTED.

Whereas I've applied a per-document time of 0.2s. So for 2500 docs, the method will wait for 0.2s * 2500 => 500s.